lib/mix/lib/mix/tasks/xref.ex

```diff
@@ -164,7 +166,11 @@ defmodule Mix.Tasks.Xref do
 
   defp unreachable(pair_fun) do
     excludes = excludes()
-    each_source_entries(&source_warnings(&1, excludes), pair_fun)
+
+    for source(source: file) = source <- sources(),
+        entries = source_warnings(source, excludes),
+        entries != [],
+        do: pair_fun.(file, entries)
```

Comment:
I think we will need to revisit this approach. If you look at `warnings/2`, you will see it has to return something like `{file, warnings}` but the warnings now can belong to many files. We will probably need to first get all warnings in a structure like `{file, line, {module, function, arity}}` and then let each part of `xref` group it as they find necessary. For example, callers may want to group by `{module, function, arity}`. The warnings will want to group by `file` and so on.

 I think by then we will no longer need `pair_fun` and instead we will end-up doing something like `unreachable() |> group_by_x |> print`.

Thoughts?